File: ios/App/App/ModbusTcpPlugin.swift
import Foundation
import Capacitor
import Network

@objc(ModbusTcpPlugin)
public class ModbusTcpPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ModbusTcpPlugin"
    public let jsName = "ModbusTcp"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "readRegisters", returnType: CAPPluginReturnPromise)
    ]

    private static let maxRegsPerRead = 125
    private static let timeoutSeconds: TimeInterval = 5

    @objc func readRegisters(_ call: CAPPluginCall) {
        guard let hostStr = call.getString("host"), !hostStr.isEmpty else {
            call.reject("host is required")
            return
        }

        let unitId = UInt8(call.getInt("unitId") ?? 1)
        let fc = UInt8(call.getInt("functionCode") ?? 0x04)
        let startAddr = call.getInt("startAddress") ?? 0
        let endAddr = call.getInt("endAddress") ?? 0
        let totalCount = endAddr - startAddr + 1

        guard totalCount > 0 && totalCount <= 0xFFFF else {
            call.reject("Invalid address range")
            return
        }

        // Parse host:port
        let host: String
        let port: UInt16
        if hostStr.contains(":") {
            let parts = hostStr.split(separator: ":", maxSplits: 1)
            host = String(parts[0])
            port = UInt16(parts[1]) ?? 502
        } else {
            host = hostStr
            port = 502
        }

        DispatchQueue.global(qos: .userInitiated).async {
            self.performRead(host: host, port: port, unitId: unitId, fc: fc,
                           startAddr: startAddr, totalCount: totalCount, call: call)
        }
    }

    private func performRead(host: String, port: UInt16, unitId: UInt8, fc: UInt8,
                            startAddr: Int, totalCount: Int, call: CAPPluginCall) {
        let nwHost = NWEndpoint.Host(host)
        let nwPort = NWEndpoint.Port(rawValue: port)!
        let connection = NWConnection(host: nwHost, port: nwPort, using: .tcp)

        let semaphore = DispatchSemaphore(value: 0)
        var connectError: Error?

        connection.stateUpdateHandler = { state in
            switch state {
            case .ready:
                semaphore.signal()
            case .failed(let error):
                connectError = error
                semaphore.signal()
            case .cancelled:
                connectError = NSError(domain: "ModbusTcp", code: -1,
                    userInfo: [NSLocalizedDescriptionKey: "Connection cancelled"])
                semaphore.signal()
            default:
                break
            }
        }

        connection.start(queue: .global(qos: .userInitiated))

        let connectResult = semaphore.wait(timeout: .now() + Self.timeoutSeconds)
        if connectResult == .timedOut {
            connection.cancel()
            call.reject("Modbus connect timeout to \(host):\(port)")
            return
        }
        if let err = connectError {
            connection.cancel()
            call.reject("Modbus connect error: \(err.localizedDescription)")
            return
        }

        var allRegisters: [Int] = []
        var offset = 0

        while offset < totalCount {
            let chunkCount = min(Self.maxRegsPerRead, totalCount - offset)
            let chunkAddr = startAddr + offset

            do {
                let regs = try readChunk(connection: connection, unitId: unitId, fc: fc,
                                        addr: chunkAddr, count: chunkCount)
                allRegisters.append(contentsOf: regs)
                offset += chunkCount
            } catch {
                connection.cancel()
                call.reject("Modbus error: \(error.localizedDescription)")
                return
            }
        }

        connection.cancel()

        var result = JSObject()
        result["registers"] = allRegisters
        call.resolve(result)
    }

    private func readChunk(connection: NWConnection, unitId: UInt8, fc: UInt8,
                          addr: Int, count: Int) throws -> [Int] {
        let transactionId = UInt16.random(in: 0...0xFFFE)

        // Build request (12 bytes)
        var request = Data(count: 12)
        request[0] = UInt8(transactionId >> 8)
        request[1] = UInt8(transactionId & 0xFF)
        request[2] = 0 // protocol ID
        request[3] = 0
        request[4] = 0 // length high
        request[5] = 6 // length low
        request[6] = unitId
        request[7] = fc
        request[8] = UInt8(addr >> 8)
        request[9] = UInt8(addr & 0xFF)
        request[10] = UInt8(count >> 8)
        request[11] = UInt8(count & 0xFF)

        // Send
        let sendSem = DispatchSemaphore(value: 0)
        var sendError: Error?
        connection.send(content: request, completion: .contentProcessed { error in
            sendError = error
            sendSem.signal()
        })
        sendSem.wait()
        if let err = sendError {
            throw err
        }

        // Receive MBAP header (7 bytes)
        let headerData = try receiveExact(connection: connection, length: 7)
        let respLength = (Int(headerData[4]) << 8) | Int(headerData[5])

        // Receive remaining (respLength - 1 for unitId)
        let pduData = try receiveExact(connection: connection, length: respLength - 1)

        let respFc = pduData[0]
        if (respFc & 0x80) != 0 {
            let errorCode = pduData[1]
            throw NSError(domain: "ModbusTcp", code: Int(errorCode),
                userInfo: [NSLocalizedDescriptionKey:
                    "Modbus exception at 0x\(String(addr, radix: 16)): FC=0x\(String(respFc, radix: 16)) code=\(errorCode)"])
        }

        let byteCount = Int(pduData[1])
        var registers: [Int] = []
        for i in stride(from: 0, to: byteCount, by: 2) {
            let val = (Int(pduData[2 + i]) << 8) | Int(pduData[2 + i + 1])
            registers.append(val)
        }

        return registers
    }

    private func receiveExact(connection: NWConnection, length: Int) throws -> Data {
        let sem = DispatchSemaphore(value: 0)
        var resultData: Data?
        var resultError: Error?

        connection.receive(minimumIncompleteLength: length, maximumLength: length) { data, _, _, error in
            resultData = data
            resultError = error
            sem.signal()
        }

        let waitResult = sem.wait(timeout: .now() + Self.timeoutSeconds)
        if waitResult == .timedOut {
            throw NSError(domain: "ModbusTcp", code: -2,
                userInfo: [NSLocalizedDescriptionKey: "Modbus read timeout"])
        }
        if let err = resultError {
            throw err
        }
        guard let data = resultData, data.count >= length else {
            throw NSError(domain: "ModbusTcp", code: -3,
                userInfo: [NSLocalizedDescriptionKey: "Incomplete Modbus response: expected \(length) bytes"])
        }
        return data
    }
}

File: android/app/src/main/java/com/github/sranka/solax_fve_realtime_app/ModbusTcpPlugin.java
package com.github.sranka.solax_fve_realtime_app;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

@CapacitorPlugin(name = "ModbusTcp")
public class ModbusTcpPlugin extends Plugin {

    private static final int MAX_REGS_PER_READ = 125;
    private static final int CONNECT_TIMEOUT_MS = 5000;
    private static final int READ_TIMEOUT_MS = 5000;
    private final Random random = new Random();

    @PluginMethod
    public void readRegisters(PluginCall call) {
        String hostStr = call.getString("host", "");
        int unitId = call.getInt("unitId", 1);
        int fc = call.getInt("functionCode", 0x04);
        int startAddr = call.getInt("startAddress", 0);
        int endAddr = call.getInt("endAddress", 0);

        if (hostStr == null || hostStr.isEmpty()) {
            call.reject("host is required");
            return;
        }

        // Parse host:port
        String host;
        int port;
        if (hostStr.contains(":")) {
            String[] parts = hostStr.split(":", 2);
            host = parts[0];
            try {
                port = Integer.parseInt(parts[1]);
            } catch (NumberFormatException e) {
                call.reject("Invalid port: " + parts[1]);
                return;
            }
        } else {
            host = hostStr;
            port = 502;
        }

        int totalCount = endAddr - startAddr + 1;
        if (totalCount <= 0 || totalCount > 0xFFFF) {
            call.reject("Invalid address range");
            return;
        }

        // Run on background thread
        final String fHost = host;
        final int fPort = port;
        getBridge().executeOnMainThread(() -> {
            new Thread(() -> {
                try {
                    List<Integer> allRegisters = readAllRegisters(fHost, fPort, unitId, fc, startAddr, totalCount);
                    JSObject result = new JSObject();
                    JSArray regsArray = new JSArray();
                    for (int val : allRegisters) {
                        regsArray.put(val);
                    }
                    result.put("registers", regsArray);
                    call.resolve(result);
                } catch (Exception e) {
                    call.reject("Modbus error: " + e.getMessage());
                }
            }).start();
        });
    }

    private List<Integer> readAllRegisters(String host, int port, int unitId, int fc, int startAddr, int totalCount) throws IOException {
        List<Integer> allRegisters = new ArrayList<>();

        Socket socket = new Socket();
        try {
            socket.connect(new InetSocketAddress(host, port), CONNECT_TIMEOUT_MS);
            socket.setSoTimeout(READ_TIMEOUT_MS);

            DataOutputStream out = new DataOutputStream(socket.getOutputStream());
            DataInputStream in = new DataInputStream(socket.getInputStream());

            int offset = 0;
            while (offset < totalCount) {
                int chunkCount = Math.min(MAX_REGS_PER_READ, totalCount - offset);
                int chunkAddr = startAddr + offset;

                // Build MBAP header + PDU (12 bytes)
                int transactionId = random.nextInt(0xFFFF);
                byte[] request = new byte[12];
                request[0] = (byte) (transactionId >> 8);
                request[1] = (byte) (transactionId & 0xFF);
                request[2] = 0; // protocol ID high
                request[3] = 0; // protocol ID low
                request[4] = 0; // length high
                request[5] = 6; // length low (unitId + PDU = 6)
                request[6] = (byte) unitId;
                request[7] = (byte) fc;
                request[8] = (byte) (chunkAddr >> 8);
                request[9] = (byte) (chunkAddr & 0xFF);
                request[10] = (byte) (chunkCount >> 8);
                request[11] = (byte) (chunkCount & 0xFF);

                out.write(request);
                out.flush();

                // Read MBAP header (7 bytes)
                byte[] mbapHeader = new byte[7];
                in.readFully(mbapHeader);

                int respLength = ((mbapHeader[4] & 0xFF) << 8) | (mbapHeader[5] & 0xFF);

                // Read remaining PDU
                byte[] pdu = new byte[respLength - 1]; // -1 for unitId already in header
                in.readFully(pdu);

                int respFc = pdu[0] & 0xFF;
                if ((respFc & 0x80) != 0) {
                    int errorCode = pdu[1] & 0xFF;
                    throw new IOException("Modbus exception at 0x" +
                        Integer.toHexString(chunkAddr) + ": FC=0x" +
                        Integer.toHexString(respFc) + " code=" + errorCode);
                }

                int byteCount = pdu[1] & 0xFF;
                for (int i = 0; i < byteCount; i += 2) {
                    int val = ((pdu[2 + i] & 0xFF) << 8) | (pdu[2 + i + 1] & 0xFF);
                    allRegisters.add(val);
                }

                offset += chunkCount;
            }
        } finally {
            try { socket.close(); } catch (IOException ignored) {}
        }

        return allRegisters;
    }
}
